src/packet_analysis/protocol/ip/IP.cc

```diff
@@ -1,37 +1,275 @@
 // See the file "COPYING" in the main distribution directory for copyright.
 
 #include "IP.h"
-#include "NetVar.h"
+#include "zeek/NetVar.h"
+#include "zeek/IP.h"
+#include "zeek/Discard.h"
+#include "zeek/PacketFilter.h"
+#include "zeek/Sessions.h"
+#include "zeek/RunState.h"
+#include "zeek/Frag.h"
+#include "zeek/Event.h"
+#include "zeek/TunnelEncapsulation.h"
 
 using namespace zeek::packet_analysis::IP;
 
 IPAnalyzer::IPAnalyzer()
 	: zeek::packet_analysis::Analyzer("IP")
+	{
+	discarder = new detail::Discarder();
+	if ( ! discarder->IsActive() )
+		{
+		delete discarder;
+		discarder = nullptr;
+		}
+	}
+
+IPAnalyzer::~IPAnalyzer()
 	{
 	}
 
 bool IPAnalyzer::AnalyzePacket(size_t len, const uint8_t* data, Packet* packet)
 	{
-	// Assume we're pointing at IP. Just figure out which version.
-	if ( sizeof(struct ip) >= len )
+	EncapsulationStack* encapsulation = packet->encap;
+
+	// Check to make sure we have enough data left for an IP header to be here. Note we only
+	// check ipv4 here. We'll check ipv6 later once we determine we have an ipv6 header.
+	if ( len < sizeof(struct ip) )
 		{
-		packet->Weird("packet_analyzer_truncated_header");
+		packet->Weird("truncated_IP");
 		return false;
 		}
 
+	// TODO: i feel like this could be generated as we move along the header hierarchy.
+	// TODO: the sessions code expects that the header size does not include the ip header. Should
+	// this change?
+	packet->hdr_size = static_cast<int32_t>(data - packet->data);
+
+	// Cast the current data pointer to an IP header pointer so we can use it to get some
+	// data about the header.
 	auto ip = (const struct ip *)data;
 	uint32_t protocol = ip->ip_v;
 
-	auto inner_analyzer = Lookup(protocol);
-	if ( inner_analyzer == nullptr )
+	// This is a unique pointer because of the mass of early returns from this method.
+	IP_Hdr* ip_hdr = nullptr;
+	if ( protocol == 4 )
 		{
-		DBG_LOG(DBG_PACKET_ANALYSIS, "Analysis in %s failed, could not find analyzer for identifier %#x.",
-				GetAnalyzerName(), protocol);
-		packet->Weird("no_suitable_analyzer_found");
+		ip_hdr = new IP_Hdr(ip, false);
+		packet->l3_proto = L3_IPV4;
+		}
+	else if ( protocol == 6 )
+		{
+		if ( len < sizeof(struct ip6_hdr) )
+			{
+			packet->Weird("truncated_IP");
+			return false;
+			}
+
+		ip_hdr = new IP_Hdr((const struct ip6_hdr*) data, false, len);
+		packet->l3_proto = L3_IPV6;
+		}
+	else
+		{
+		packet->Weird("unknown_ip_version");
 		return false;
 		}
 
-	DBG_LOG(DBG_PACKET_ANALYSIS, "Analysis in %s succeeded, next layer identifier is %#x.",
```

Comment:
Did we lose this? I have found this message helpful to have, as it doesn't seem there's other debug messages recording the progress of packet analysis.

Reply:
That should be handled entirely by `Analyzer::ForwardPacket` now. It is, however, lost when we pass it on to the Session analyzer. I can add it back in there though.

Reply:
adding there sounds good